src/runtimes/callback.rs

```diff
@@ -0,0 +1,429 @@
+                    req,
+                    rx,
+                    eval_ctx,
+                )
+            }
+            "namespaces" => {
+                let req = CallbackRequestType::KubernetesListResourceAll {
+                    api_version: "v1".to_string(),
+                    kind: "Namespace".to_string(),
+                    label_selector: None,
+                    field_selector: None,
+                };
+
+                warn!(
+                    eval_ctx.policy_id,
+                    ?req,
+                    "Usage of deprecated `ClusterContext`"
+                );
+                let (tx, rx) = oneshot::channel::<Result<CallbackResponse>>();
+                let req = CallbackRequest {
+                    request: req,
+                    response_channel: tx,
+                };
+                send_request_and_wait_for_response(
+                    &eval_ctx.policy_id,
+                    binding,
+                    operation,
+                    req,
+                    rx,
+                    eval_ctx,
+                )
+            }
+            "services" => {
+                let req = CallbackRequestType::KubernetesListResourceAll {
+                    api_version: "v1".to_string(),
+                    kind: "Service".to_string(),
+                    label_selector: None,
+                    field_selector: None,
+                };
+
+                warn!(
+                    eval_ctx.policy_id,
+                    ?req,
+                    "Usage of deprecated `ClusterContext`"
+                );
+                let (tx, rx) = oneshot::channel::<Result<CallbackResponse>>();
+                let req = CallbackRequest {
+                    request: req,
+                    response_channel: tx,
+                };
+                send_request_and_wait_for_response(
+                    &eval_ctx.policy_id,
+                    binding,
+                    operation,
+                    req,
+                    rx,
+                    eval_ctx,
+                )
+            }
+            _ => {
+                error!("unknown namespace: {}", namespace);
+                Err(format!("unknown namespace: {namespace}").into())
+            }
+        },
+        _ => {
+            error!("unknown binding: {}", binding);
+            Err(format!("unknown binding: {binding}").into())
+        }
+    }
+}
+
+fn send_request_and_wait_for_response(
+    policy_id: &str,
+    binding: &str,
+    operation: &str,
+    req: CallbackRequest,
+    mut rx: Receiver<Result<CallbackResponse>>,
+    eval_ctx: &EvaluationContext,
+) -> Result<Vec<u8>, Box<dyn std::error::Error + Send + Sync>> {
+    let cb_channel: mpsc::Sender<CallbackRequest> =
+        if let Some(c) = eval_ctx.callback_channel.clone() {
+            Ok(c)
+        } else {
+            error!(
+                policy_id,
+                binding, operation, "Cannot process waPC request: callback channel not provided"
```

Comment:
```suggestion
                binding, operation, "Cannot process request: callback channel not provided"
```

These requests may be waPC or Wasi (same for other err messages).